### DISCUSSION TITLE
Rename Dial to DialStream and DialPacket

### DISCUSSION THREAD

**Discussion by fortuna** [OutlineFoundation/outline-sdk#150] [API Proposals and Ideas]
Currently StreamDialers and PacketDialers both have a Dial method. That means an object can't be both, but there are many situations where we want to be able to dial either streams or packets (e.g. a DNS resolver, a SOCKS proxy,...).

I propose we rename the Dial method to DialStream and DialPacket, so that an object can implement both.

The same applies to Endpoints. We need ConnectStream and ConnectPacket.

This is obviously a breaking change, but we don't yet promise API stability (we are still v0).

/cc @jyyi1

**Comment by fortuna** [comment]
One potential downside to this is that it would prevent the use of Generics. Generics has the potential to reduce code duplication, and I played with it a bit in a [fork](https://github.com/Jigsaw-Code/outline-sdk/pull/19). Generics in Go are pretty bad though, and I ran into a lot of issues with the type inference and dealing with it was even more of a pain.

**Comment by jyyi1** [comment]
I'm OK with the duplicated logic, but I would prefer to do it early than late because it is a big breaking change. Another option is to add new interfaces, and gradually move old implementation to the new one.

**Comment by fortuna** [comment]
I agree with doing it sooner rather than later. Though there are big PRs still under way to merge first.

I agree with the adding new interfaces, if we were promising a stable API. The point of keeping it in v0 for now is to enable us to make breaking changes before we commit to a stable API, as we learn from real world usage. Having two competing Dial methods will be a pain, like it is with the standard library Dial/DialContext.

**Comment by fortuna** [comment]
@jyyi1  PR implementing the proposal: https://github.com/Jigsaw-Code/outline-sdk/pull/165

**Comment by fortuna** [comment]
Implemented and released: https://github.com/Jigsaw-Code/outline-sdk/releases/tag/v0.0.12
